File: sdk/python/examples/langgraph/21_subgraph.py
# Copyright (c) 2025 Agentspan
# Licensed under the MIT License. See LICENSE file in the project root for details.

"""Subgraph — composing analysis sub-tasks as tools within a parent agent.

Demonstrates:
    - Each analysis sub-task (sentiment, keywords, summary) is a tool
    - A report-building tool combines all sub-task outputs
    - The LLM orchestrates the full pipeline server-side via tool calls
    - Practical use case: document processing pipeline with nested analysis

Requirements:
    - AGENTSPAN_SERVER_URL=http://localhost:8080/api
    - OPENAI_API_KEY for ChatOpenAI
"""

from langchain_core.tools import tool
from langchain_core.prompts import ChatPromptTemplate
from langchain_openai import ChatOpenAI
from agentspan.agents.langchain import create_agent
from agentspan.agents import AgentRuntime

llm = ChatOpenAI(model="gpt-4o-mini", temperature=0)


@tool
def analyze_sentiment(text: str) -> str:
    """Classify the sentiment of the given text.

    Returns one of: positive, negative, or neutral.

    Args:
        text: Text to classify sentiment for.
    """
    prompt = ChatPromptTemplate.from_messages([
        ("system", "Classify the sentiment of the text. Return ONLY: positive, negative, or neutral."),
        ("human", "{text}"),
    ])
    chain = prompt | llm
    response = chain.invoke({"text": text})
    return response.content.strip().lower()


@tool
def extract_keywords(text: str) -> str:
    """Extract 3-5 keywords from the text as a comma-separated list.

    Args:
        text: Text to extract keywords from.
    """
    prompt = ChatPromptTemplate.from_messages([
        ("system", "Extract 3-5 keywords from the text. Return a comma-separated list only."),
        ("human", "{text}"),
    ])
    chain = prompt | llm
    response = chain.invoke({"text": text})
    return response.content.strip()


@tool
def summarize_text(text: str) -> str:
    """Summarize the given text in one concise sentence.

    Args:
        text: Text to summarize.
    """
    prompt = ChatPromptTemplate.from_messages([
        ("system", "Summarize this text in one sentence."),
        ("human", "{text}"),
    ])
    chain = prompt | llm
    response = chain.invoke({"text": text})
    return response.content.strip()


@tool
def build_report(sentiment: str, keywords: str, summary: str) -> str:
    """Combine analysis results into a formatted document analysis report.

    Args:
        sentiment: The sentiment classification result.
        keywords: The extracted keywords (comma-separated).
        summary: The one-sentence summary.
    """
    return (
        f"Document Analysis Report\n"
        f"========================\n"
        f"Sentiment:  {sentiment}\n"
        f"Keywords:   {keywords}\n"
        f"Summary:    {summary}\n"
    )


PIPELINE_SYSTEM = """You are a document analysis orchestrator.

For each document:
1. Call analyze_sentiment with the document text
2. Call extract_keywords with the document text
3. Call summarize_text with the document text
4. Call build_report with the sentiment, keywords, and summary from steps 1-3
5. Return the report to the user

Always complete all four analysis steps before building the report.
"""

graph = create_agent(
    llm,
    tools=[analyze_sentiment, extract_keywords, summarize_text, build_report],
    name="document_pipeline_with_subgraph",
    system_prompt=PIPELINE_SYSTEM,
)

if __name__ == "__main__":
    sample_doc = (
        "LangGraph makes it easy to build stateful, multi-actor applications with LLMs. "
        "The framework provides first-class support for persistence, streaming, and human-in-the-loop "
        "workflows. Developers love its flexibility and the ability to compose complex pipelines "
        "using simple Python functions."
    )

    with AgentRuntime() as runtime:
        result = runtime.run(graph, sample_doc)
        print(f"Status: {result.status}")
        result.print_result()

File: sdk/typescript/examples/langgraph/21-subgraph.ts
/**
 * Subgraph -- composing graphs within graphs.
 *
 * Demonstrates:
 *   - Building a nested subgraph for a specific subtask
 *   - Connecting a subgraph as a node in a parent graph
 *   - Passing state between parent graph and subgraph
 *   - Practical use case: document processing pipeline with a nested analysis subgraph
 */

import { StateGraph, START, END, Annotation } from '@langchain/langgraph';
import { ChatOpenAI } from '@langchain/openai';
import { HumanMessage, SystemMessage } from '@langchain/core/messages';
import { AgentRuntime } from '../../src/index.js';

const llm = new ChatOpenAI({ model: 'gpt-4o-mini', temperature: 0 });

// ---------------------------------------------------------------------------
// Subgraph state schema
// ---------------------------------------------------------------------------
const AnalysisState = Annotation.Root({
  text: Annotation<string>({
    reducer: (_prev: string, next: string) => next ?? _prev,
    default: () => '',
  }),
  sentiment: Annotation<string>({
    reducer: (_prev: string, next: string) => next ?? _prev,
    default: () => '',
  }),
  keywords: Annotation<string[]>({
    reducer: (_prev: string[], next: string[]) => next ?? _prev,
    default: () => [],
  }),
  summary: Annotation<string>({
    reducer: (_prev: string, next: string) => next ?? _prev,
    default: () => '',
  }),
});

type AnalysisStateType = typeof AnalysisState.State;

// ---------------------------------------------------------------------------
// Subgraph node functions
// ---------------------------------------------------------------------------
async function analyzeSentiment(state: AnalysisStateType): Promise<Partial<AnalysisStateType>> {
  const response = await llm.invoke([
    new SystemMessage(
      'Classify the sentiment of the text. Return ONLY: positive, negative, or neutral.',
    ),
    new HumanMessage(state.text),
  ]);
  return { sentiment: (response.content as string).trim().toLowerCase() };
}

async function extractKeywords(state: AnalysisStateType): Promise<Partial<AnalysisStateType>> {
  const response = await llm.invoke([
    new SystemMessage(
      'Extract 3-5 keywords from the text. Return a comma-separated list only.',
    ),
    new HumanMessage(state.text),
  ]);
  const keywords = (response.content as string).split(',').map((k) => k.trim());
  return { keywords };
}

async function summarizeText(state: AnalysisStateType): Promise<Partial<AnalysisStateType>> {
  const response = await llm.invoke([
    new SystemMessage('Summarize this text in one sentence.'),
    new HumanMessage(state.text),
  ]);
  return { summary: (response.content as string).trim() };
}

// ---------------------------------------------------------------------------
// Build the subgraph
// ---------------------------------------------------------------------------
const analysisBuilder = new StateGraph(AnalysisState);
analysisBuilder.addNode('sentiment_node', analyzeSentiment);
analysisBuilder.addNode('keywords_node', extractKeywords);
analysisBuilder.addNode('summarize', summarizeText);
analysisBuilder.addEdge(START, 'sentiment_node');
analysisBuilder.addEdge('sentiment_node', 'keywords_node');
analysisBuilder.addEdge('keywords_node', 'summarize');
analysisBuilder.addEdge('summarize', END);
const analysisSubgraph = analysisBuilder.compile();

// ---------------------------------------------------------------------------
// Parent graph state schema
// ---------------------------------------------------------------------------
const DocumentState = Annotation.Root({
  document: Annotation<string>({
    reducer: (_prev: string, next: string) => next ?? _prev,
    default: () => '',
  }),
  analysis_text: Annotation<string>({
    reducer: (_prev: string, next: string) => next ?? _prev,
    default: () => '',
  }),
  sentiment: Annotation<string>({
    reducer: (_prev: string, next: string) => next ?? _prev,
    default: () => '',
  }),
  keywords: Annotation<string[]>({
    reducer: (_prev: string[], next: string[]) => next ?? _prev,
    default: () => [],
  }),
  summary: Annotation<string>({
    reducer: (_prev: string, next: string) => next ?? _prev,
    default: () => '',
  }),
  report: Annotation<string>({
    reducer: (_prev: string, next: string) => next ?? _prev,
    default: () => '',
  }),
});

type DocumentStateType = typeof DocumentState.State;

// ---------------------------------------------------------------------------
// Parent graph node functions
// ---------------------------------------------------------------------------
function prepare(state: DocumentStateType): Partial<DocumentStateType> {
  // Use the whole document as the analysis text
  return { analysis_text: state.document };
}

async function runAnalysis(state: DocumentStateType): Promise<Partial<DocumentStateType>> {
  const result = await analysisSubgraph.invoke({ text: state.analysis_text });
  return {
    sentiment: result.sentiment ?? '',
    keywords: result.keywords ?? [],
    summary: result.summary ?? '',
  };
}

function buildReport(state: DocumentStateType): Partial<DocumentStateType> {
  const keywordsStr = (state.keywords ?? []).join(', ');
  const report =
    'Document Analysis Report\n' +
    '========================\n' +
    `Sentiment:  ${state.sentiment ?? 'unknown'}\n` +
    `Keywords:   ${keywordsStr}\n` +
    `Summary:    ${state.summary ?? ''}`;
  return { report };
}

// ---------------------------------------------------------------------------
// Build the parent graph
// ---------------------------------------------------------------------------
const parentBuilder = new StateGraph(DocumentState);
parentBuilder.addNode('prepare', prepare);
parentBuilder.addNode('analysis', runAnalysis);
parentBuilder.addNode('build_report', buildReport);
parentBuilder.addEdge(START, 'prepare');
parentBuilder.addEdge('prepare', 'analysis');
parentBuilder.addEdge('analysis', 'build_report');
parentBuilder.addEdge('build_report', END);

const graph = parentBuilder.compile();

// Add agentspan metadata for extraction
(graph as any)._agentspan = {
  model: 'openai/gpt-4o-mini',
  tools: [],
  framework: 'langgraph',
};

const PROMPT =
  'LangGraph makes it easy to build stateful, multi-actor applications with LLMs. ' +
  'The framework provides first-class support for persistence, streaming, and human-in-the-loop ' +
  'workflows. Developers love its flexibility and the ability to compose complex pipelines ' +
  'using simple Python functions.';

// ---------------------------------------------------------------------------
// Run on agentspan
// ---------------------------------------------------------------------------
async function main() {
  const runtime = new AgentRuntime();
  try {
    const result = await runtime.run(graph, PROMPT);
    console.log('Status:', result.status);
    result.printResult();
  } finally {
    await runtime.shutdown();
  }
}

main().catch(console.error);
